fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,784 @@
+    deleted_keys: BTreeSet<K>,
+}
+
+impl<K, V> AsRef<BTreeMap<K, V>> for FolderStorage<K, V>
+where
+    K: Ord,
+{
+    fn as_ref(&self) -> &BTreeMap<K, V> {
+        &self.data
+    }
+}
+
+impl<K, V> FolderStorage<K, V>
+where
+    K: Ord
+        + Clone
+        + serde::Serialize
+        + serde::de::DeserializeOwned
+        + std::str::FromStr
+        + std::fmt::Display,
+    V: Clone + serde::Serialize + serde::de::DeserializeOwned + Monoid<V>,
+{
+    /// Create a new folder storage with a diagnostic label and directory path
+    /// The storage will be initialized using the disk data, if the path exists
+    ///
+    /// Note: if the folder storage already exists, the data will be read from the folder
+    /// without overwriting it.
+    pub fn new(label: String, path: &Path) -> Result<Self> {
+        let mut storage = Self {
+            label,
+            path: PathBuf::from(path),
+            modified: BTreeMap::new(),
+            written_to_disk: BTreeMap::new(),
+            data: BTreeMap::new(),
+            deleted_keys: BTreeSet::new(),
+        };
+
+        if Path::exists(path) {
+            storage.read_fs()?;
+        }
+
+        Ok(storage)
+    }
+
+    /// Load mapping from folder storage
+    fn load_fs_data(&mut self) -> Result<()> {
+        if !self.path.exists() {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                "Folder does not exist".to_owned(),
+            ));
+        }
+
+        if !self.path.is_dir() {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                "Path is not a directory".to_owned(),
+            ));
+        }
+
+        let mut data = BTreeMap::new();
+
+        for entry in fs::read_dir(&self.path)? {
+            let entry = entry?;
+            let path = entry.path();
+            if path.is_file()
+                && path
+                    .extension()
+                    .map_or(false, |ext| ext == "json")
+            {
+                let key: K = self.extract_key_from_file_path(&path)?;
+                let file = File::open(&path)?;
+                let value: V =
+                    serde_json::from_reader(file).map_err(|err| {
+                        ArklibError::Storage(
+                            self.label.clone(),
+                            err.to_string(),
+                        )
+                    })?;
+
+                data.insert(key.clone(), value);
+
+                if let Ok(metadata) = fs::metadata(&path) {
+                    if let Ok(modified) = metadata.modified() {
+                        self.written_to_disk.insert(key.clone(), modified);
+                        self.modified.insert(key, modified);
+                    }
+                }
+            }
+        }
+
+        self.data = data;
+        Ok(())
+    }
+    /// Resolve differences between memory and disk data
+    fn resolve_divergence(&mut self) -> Result<()> {
```

Comment:
I'm sorry, but could you explain what this method does?

Reply:
It is just an enhanced version of 'merge_from' for resolving divergence. The method considers timestamps to detect keys where we need to use monoids to resolve conflicts.


Added this comment here:
```rs
    /// Resolves discrepancies between in-memory data and disk data by combining or
    /// overwriting values based on which version is more recent, ensuring consistency.

```